### DISCUSSION TITLE
MaxAge Setting per JetStream is very aggressive

### DISCUSSION THREAD

**Discussion by stefanLeo** [nats-io/nats-server#5136] [General]
I just noticed within my testsetup that memstores expireMsgs is called basically all around the clock stopping the store with the lock, deleting 2-3 msgs and then starting all over again in a few micro secs...

This has a performance hit and I would like to discuss if this aggressiveness is really needed. 
My use case: I have a quite steady rate of msgs and I have set the max age to something like 30 mins. 
On my stream I have ~100 subjects where there is a lot of traffic on some, but almost none on others.  I also have set a per subj msg limit. I cannot tell beforehand which subjects will be the ones with high and with low traffic in the real world as this is depending on user behaviour at the end of the day. 
This setup causes the deleted count to go up as discussed in another question.

Goal was to find a sweet spot with config between maxage, maxbytes and maxmsgs per subject to reduce those delete messages. 
MaxAge was added now to get rid of such msgs stuck in subjects with almost no new messages sent. 
My expectation based on the docu was: If I set this to something like 30 mins, then this limit is evaluated like once every few mins or so. I was not aware that these 30mins are taken literally as exact as a nanosec level. 

Ideas / Proposals:
- Let the stream creator decide how often such expireMsgs calls are done. As people might not need it to be "super accurate".
- Relax the times it runs in general to not run it more often than X times per minute/second ...  i.e., some sort of min firetime
- Filter out msgs that are too old when a new consumer subscribes to achieve accuracy on the interface level, but do not enforce real-time deletion on the store and then run the actual deletion on the store like half the value provided by the creator (15 mins in my example) and at least once an hour or so...
-  align expireMsgs to the filestore's expireMsgs which locks only inside the loop, which should give it a higher chance to do continuous deletions in one go ... 

Not sure though if running it at a lesser pace effectively deleting more messages in one go has other performance implications?

**Comment by derekcollison** [comment]
The system should only fire the timer based on when it thinks the next message will expire. We could soften that a bit and round up to say seconds, but would probably not go beyond that since had memory repercussions.

If we decide to go forward with a second minimum timer value I would be ok with that.

**Reply by stefanLeo** [reply]
OK understood. I will try a few approaches to see if they acutally help and document my findings here. 
Just an additional input: In my scenario I had 27067 runs of "expireMsgs" which expired 1 msg per run within a 30sec trace span.

**Reply by derekcollison** [reply]
Understood, you had those messages come in very quickly so the resolution is working against us in how we determine when to fire the next timer to check..

**Reply by stefanLeo** [reply]
Summary of my tests: This is a classic throughput vs. latency topic and you have to pick your poison. 
Todays implementation focuses on latency mostly. If there is a burst of incoming messsages roughly withtin the same period of time this then leads to big outages in the insert performance once those messages need to get expired. 
On the other hand during steady ingest, the routine runs very often, deleting just 1 or 2 msgs and then being restarted a few usec later. This however seems to be no major issue, expcet that it impacts storemsgs performance overall but only by say a few usecs.. 

I came to the conclusion that this is maybe the sweet spot:

```
func (ms *memStore) expireMsgs() {
	for {
		ms.mu.Lock()
		bDone := false
		for i := 1; i < 100; i++ {
			sm, ok := ms.msgs[ms.state.FirstSeq]
			minAge := time.Now().UnixNano() - int64(ms.cfg.MaxAge)
			if ok && sm.ts <= minAge {
				ms.deleteFirstMsgOrPanic()
			} else {
				if len(ms.msgs) == 0 {
					if ms.ageChk != nil {
						ms.ageChk.Stop()
						ms.ageChk = nil
					}
				} else {
					fireIn := min(time.Duration(sm.ts-minAge), time.Second)
					if sm == nil {
						fireIn = ms.cfg.MaxAge
					}
					if ms.ageChk != nil {
						ms.ageChk.Reset(fireIn)
					} else {
						ms.ageChk = time.AfterFunc(fireIn, ms.expireMsgs)
					}
				}
				bDone = true
				break
			}
		}
		ms.mu.Unlock()
		if bDone {
			return
		}
	}
}
```

**Reply by stefanLeo** [reply]
just out of curiosity: Is there any better approach to evict old messages (across different subjects) but not using maxAge at all?

**Reply by derekcollison** [reply]
If there are per subject limits we evict on those thresholds. For age based, I will look to have a minimum like you proposed above.

**Reply by stefanLeo** [reply]
Pls. also consider the inner loop idea as it releases the lock and can improve latency in certain high volume insert situations especially as the minage is always adapted...  in one of my tests I managed to block new insertions for ~1 min as expireMsgs did not stop running for that long and hence never released the lock...

**Reply by derekcollison** [reply]
That is what we do in filestore so I agree. Will make change.

**Reply by derekcollison** [reply]
Done.

**Comment by stefanLeo** [comment]
Thank you ... that was quick 🚀

**Reply by derekcollison** [reply]
We aim to please ;)
